src/node.rs

```diff
@@ -0,0 +1,147 @@
+//! The node that makes up queues.
+
+use crate::inner::Inner;
+use crate::list::{Entry, List, State};
+use crate::sync::atomic::AtomicPtr;
+use crate::{Notify, NotifyKind, Task};
+
+use alloc::boxed::Box;
+use core::ptr::{self, NonNull};
+
+/// A node in the backup queue.
+pub(crate) struct Node {
+    /// The next node in the queue.
+    next: AtomicPtr<Node>,
+
+    /// The data associated with the node.
+    data: Option<NodeData>,
+}
+
+impl From<NodeData> for Node {
+    fn from(data: NodeData) -> Self {
+        Self {
+            next: AtomicPtr::new(ptr::null_mut()),
+            data: Some(data),
+        }
+    }
+}
+
+enum NodeData {
+    /// This node is requesting to add a listener.
+    AddListener {
+        /// The pointer to the listener to add.
+        listener: NonNull<Entry>,
+    },
+
+    /// This node is notifying a listener.
+    Notify(Notify),
+
+    /// This node is removing a listener.
+    RemoveListener {
+        /// The pointer to the listener to remove.
+        listener: NonNull<Entry>,
+
+        /// Whether to propagate notifications to the next listener.
+        propagate: bool,
+    },
+
+    /// We are waiting for the mutex to lock, so they can manipulate it.
+    Waiting(Task),
+}
+
+impl Node {
+    /// Create a new listener submission entry.
+    pub(crate) fn listener() -> (Self, NonNull<Entry>) {
+        // Allocate an entry on the heap.
+        let entry = unsafe { NonNull::new_unchecked(Box::into_raw(Box::new(Entry::new()))) };
+
+        (NodeData::AddListener { listener: entry }.into(), entry)
+    }
+
+    /// Create a new notification entry.
+    pub(crate) fn notify(notify: Notify) -> Self {
+        NodeData::Notify(notify).into()
+    }
+
+    /// Create a new listener removal entry.
+    pub(crate) fn remove_listener(listener: NonNull<Entry>, propagate: bool) -> Self {
+        NodeData::RemoveListener {
+            listener,
+            propagate,
+        }
+        .into()
+    }
+
+    /// Create a new waiting entry.
+    pub(crate) fn waiting(task: Task) -> Self {
+        NodeData::Waiting(task).into()
+    }
+
+    pub(crate) fn next(&self) -> &AtomicPtr<Node> {
+        &self.next
+    }
```

Comment:
make `.next` `pub(crate)` instead, the pointer is atomic, so that we only allow `&` access here doesn't seem to matter.